jl4-core/src/L4/TypeCheck/Types.hs

```diff
@@ -292,20 +294,86 @@ resolvedType n = do
         KnownType kind _resolved _tyDecl -> setAnnResolvedKindOfResolved kind n
         KnownTerm ty _term -> setAnnResolvedTypeOfResolved ty n
         KnownTypeVariable -> setAnnResolvedKindOfResolved 0 n
+        KnownSection _ -> n -- TODO: this is probably not what we want, maybe some internal error?
 
 lookupRawNameInEnvironment :: RawName -> Check [(Unique, Name, CheckEntity)]
-lookupRawNameInEnvironment n = do
+lookupRawNameInEnvironment rn = do
+  -- NOTE: split in a list of qualifiers and the actual name
+  let (tn, qs)  = case rn of
+        QualifiedName n' qs' -> (NormalName n', NE.toList qs')
+        n -> (n, [])
+
   env <- asks (.environment)
   ei  <- asks (.entityInfo)
-  let
-    proc :: Unique -> Maybe (Unique, Name, CheckEntity)
-    proc u = (\ (o, ce) -> (u, o, ce)) <$> Map.lookup u ei
 
-    candidates :: [(Unique, Name, CheckEntity)]
-    candidates = mapMaybe proc (Map.findWithDefault [] n env)
+  let -- NOTE: from a list of qualifiers, try to build all paths through sections that may be
+      -- meant with the qualified name
+      -- The outer list describes the path, the inner list describes all possibilities at a
+      -- section level.
+      mkCandidateSections :: [Text] -> [[Section Resolved]]
+      mkCandidateSections =
+        foldl
+          (\acc q -> case Map.lookup (NormalName q) env of
+            Just us ->  mapMaybe isKnownSection us : acc
+
+            -- NOTE: this case is important because it cuts off options as soon as we run into unknown sections,
+            -- i.e. Foo.Bar.baz where Foo does exist (and contains baz) but Bar doesn't
+            Nothing -> [[]]
+          )
+          []
+
+      isKnownSection :: Unique -> Maybe (Section Resolved)
+      isKnownSection u = do
+        (_n, KnownSection s) <- Map.lookup u ei
+        pure s
+
+      -- NOTE: this is used to filter the paths through sections down to the ones that are valid.
+      -- We return the sections that are contained in existing sections
+      sectionInSection :: [[Section Resolved]] -> [Section Resolved]
+      -- NOTE: we have to treat this specially otherwise the recursive call will always
+      -- bottom out at an empty list of sections which makes the entire algorithm fail
+      sectionInSection [secs] = secs
+      sectionInSection (secs : secss) =
+        mapMaybe
+          (\sec@(MkSection _ mn _ _) -> do
+             u <- getUnique <$> mn
+             guard $ any (isTopLevelBindingInSection u) $ sectionInSection secss
+             pure sec
```

Comment:
As I said above, it would feel more natural to me to look at this in terms of `[CheckInfo]` or similar rather than a `Section Resolved`.

Intuitively, we want the *type* of the section here, not the section itself.

Reply:
I don't have the same intuition. For me this is all about scoping, nothing much to do with the types, since sections cannot have types.

Reply:
I agree that it's a bit weird since all of the constructs we had til now have had types and now it feels a bit intermingled but I think it makes sense for sections.

Reply:
Again, sections *have* types.

Reply:
alright, I'll see if that works.

Reply:
I actually need some information from the `Section` though, like its name etc. Idk how `CheckInfo` would improve the situation. I'm not seeing it (I also tried it out and I thought it would complicate things quite a lot and makes things harder to read?)